File: TypeScript/csv_utils/csv_utils_test.ts
/**
 * CSV Utilities Test Suite
 * 
 * Comprehensive tests for the CSV Utilities module.
 */

import {
  parse,
  stringify,
  isValidCsv,
  getStats,
  filterRows,
  selectColumns,
  sortBy,
  addColumn,
  removeColumn,
  renameColumn,
  mergeVertical,
  detectDelimiter,
  toJson,
  fromJson,
} from './mod';

// Test utilities
let testsPassed = 0;
let testsFailed = 0;

function test(name: string, fn: () => void): void {
  try {
    fn();
    testsPassed++;
    console.log(`✓ ${name}`);
  } catch (e) {
    testsFailed++;
    console.error(`✗ ${name}: ${e}`);
  }
}

function assertEqual(actual: unknown, expected: unknown, msg?: string): void {
  if (JSON.stringify(actual) !== JSON.stringify(expected)) {
    throw new Error(`${msg || 'Assertion failed'}: expected ${JSON.stringify(expected)}, got ${JSON.stringify(actual)}`);
  }
}

function assertTrue(value: boolean, msg?: string): void {
  if (!value) {
    throw new Error(msg || 'Expected true');
  }
}

function assertFalse(value: boolean, msg?: string): void {
  if (value) {
    throw new Error(msg || 'Expected false');
  }
}

// Test cases
console.log('Running CSV Utilities Tests...\n');

// Parse tests
test('parse simple CSV', () => {
  const csv = 'name,age\nJohn,30\nJane,25';
  const result = parse(csv);
  assertEqual(result.headers, ['name', 'age']);
  assertEqual(result.rows.length, 2);
  assertEqual(result.rows[0].name, 'John');
  assertEqual(result.rows[0].age, '30');
});

test('parse CSV with quotes', () => {
  const csv = 'name,description\n"John Doe","A person, who codes"\nJane,Developer';
  const result = parse(csv);
  assertEqual(result.rows[0].name, 'John Doe');
  assertEqual(result.rows[0].description, 'A person, who codes');
});

test('parse CSV with escaped quotes', () => {
  const csv = 'name,quote\nJohn,"He said ""Hello"""\nJane,Hi';
  const result = parse(csv);
  assertEqual(result.rows[0].quote, 'He said "Hello"');
});

test('parse CSV without headers', () => {
  const csv = 'John,30\nJane,25';
  const result = parse(csv, { header: false });
  assertEqual(result.headers, ['col1', 'col2']);
  assertEqual(result.rows[0].col1, 'John');
});

test('parse CSV with custom delimiter', () => {
  const csv = 'name;age\nJohn;30\nJane;25';
  const result = parse(csv, { delimiter: ';' });
  assertEqual(result.headers, ['name', 'age']);
  assertEqual(result.rows[0].name, 'John');
});

test('parse CSV with CRLF line endings', () => {
  const csv = 'name,age\r\nJohn,30\r\nJane,25';
  const result = parse(csv);
  assertEqual(result.rows.length, 2);
});

test('parse CSV with empty lines', () => {
  const csv = 'name,age\n\nJohn,30\n\nJane,25\n';
  const result = parse(csv);
  assertEqual(result.rows.length, 2);
});

test('parse CSV with trim disabled', () => {
  const csv = 'name,age\n John , 30 \nJane,25';
  const result = parse(csv, { trim: false });
  assertEqual(result.rows[0].name, ' John ');
});

// Stringify tests
test('stringify simple data', () => {
  const data = [{ name: 'John', age: 30 }, { name: 'Jane', age: 25 }];
  const result = stringify(data);
  assertTrue(result.includes('name,age'));
  assertTrue(result.includes('John,30'));
});

test('stringify with quotes', () => {
  const data = [{ name: 'John, Jr.', age: 30 }];
  const result = stringify(data);
  assertTrue(result.includes('"John, Jr."'));
});

test('stringify 2D array', () => {
  const data = [['name', 'age'], ['John', '30'], ['Jane', '25']];
  const result = stringify(data);
  assertTrue(result.includes('name,age'));
});

test('stringify with custom delimiter', () => {
  const data = [{ name: 'John', age: 30 }];
  const result = stringify(data, { delimiter: ';' });
  assertTrue(result.includes('name;age'));
});

test('stringify without headers', () => {
  const data = [{ name: 'John', age: 30 }];
  const result = stringify(data, { header: false });
  assertFalse(result.includes('name,age'));
});

test('stringify with alwaysQuote', () => {
  const data = [{ name: 'John', age: 30 }];
  const result = stringify(data, { alwaysQuote: true });
  assertTrue(result.includes('"John"'));
  assertTrue(result.includes('"30"'));
});

// Validation tests
test('isValidCsv with valid CSV', () => {
  assertTrue(isValidCsv('name,age\nJohn,30'));
});

test('isValidCsv with empty string', () => {
  assertTrue(isValidCsv(''));
});

// Stats tests
test('getStats', () => {
  const csv = 'name,age\nJohn,30\nJane,25';
  const data = parse(csv);
  const stats = getStats(data);
  assertEqual(stats.rowCount, 2);
  assertEqual(stats.columnCount, 2);
  assertEqual(stats.emptyCells, 0);
  assertTrue(stats.avgRowLength > 0);
});

test('getStats with empty cells', () => {
  const csv = 'name,age\nJohn,\n,25';
  const data = parse(csv);
  const stats = getStats(data);
  assertEqual(stats.emptyCells, 2);
});

// Filter tests
test('filterRows', () => {
  const csv = 'name,age\nJohn,30\nJane,25\nBob,35';
  const data = parse(csv);
  const filtered = filterRows(data, (row) => parseInt(row.age as string) > 25);
  assertEqual(filtered.rows.length, 2);
});

// Select columns tests
test('selectColumns', () => {
  const csv = 'name,age,city\nJohn,30,NYC\nJane,25,LA';
  const data = parse(csv);
  const selected = selectColumns(data, ['name', 'city']);
  assertEqual(selected.headers, ['name', 'city']);
  assertEqual(selected.rows[0].name, 'John');
  assertEqual(selected.rows[0].city, 'NYC');
});

test('selectColumns with invalid column', () => {
  const csv = 'name,age\nJohn,30';
  const data = parse(csv);
  const selected = selectColumns(data, ['name', 'invalid']);
  assertEqual(selected.headers, ['name']);
});

// Sort tests
test('sortBy numeric ascending', () => {
  const csv = 'name,age\nBob,35\nJohn,30\nJane,25';
  const data = parse(csv);
  const sorted = sortBy(data, 'age');
  assertEqual(sorted.rows[0].name, 'Jane');
  assertEqual(sorted.rows[2].name, 'Bob');
});

test('sortBy numeric descending', () => {
  const csv = 'name,age\nBob,35\nJohn,30\nJane,25';
  const data = parse(csv);
  const sorted = sortBy(data, 'age', false);
  assertEqual(sorted.rows[0].name, 'Bob');
  assertEqual(sorted.rows[2].name, 'Jane');
});

test('sortBy string', () => {
  const csv = 'name,age\nBob,35\nAlice,30\nCharlie,25';
  const data = parse(csv);
  const sorted = sortBy(data, 'name');
  assertEqual(sorted.rows[0].name, 'Alice');
  assertEqual(sorted.rows[2].name, 'Charlie');
});

test('sortBy invalid column', () => {
  const csv = 'name,age\nJohn,30';
  const data = parse(csv);
  const sorted = sortBy(data, 'invalid');
  assertEqual(sorted.rows.length, 1);
});

// Add column tests
test('addColumn', () => {
  const csv = 'name,age\nJohn,30\nJane,25';
  const data = parse(csv);
  const updated = addColumn(data, 'city', ['NYC', 'LA']);
  assertEqual(updated.headers, ['name', 'age', 'city']);
  assertEqual(updated.rows[0].city, 'NYC');
  assertEqual(updated.rows[1].city, 'LA');
});

// Remove column tests
test('removeColumn', () => {
  const csv = 'name,age,city\nJohn,30,NYC';
  const data = parse(csv);
  const updated = removeColumn(data, 'age');
  assertEqual(updated.headers, ['name', 'city']);
  assertEqual(updated.rows[0].name, 'John');
});

test('removeColumn invalid', () => {
  const csv = 'name,age\nJohn,30';
  const data = parse(csv);
  const updated = removeColumn(data, 'invalid');
  assertEqual(updated.headers, ['name', 'age']);
});

// Rename column tests
test('renameColumn', () => {
  const csv = 'name,age\nJohn,30';
  const data = parse(csv);
  const updated = renameColumn(data, 'age', 'years');
  assertEqual(updated.headers, ['name', 'years']);
  assertEqual(updated.rows[0].years, '30');
});

// Merge tests
test('mergeVertical', () => {
  const csv1 = 'name,age\nJohn,30';
  const csv2 = 'name,age\nJane,25';
  const data1 = parse(csv1);
  const data2 = parse(csv2);
  const merged = mergeVertical(data1, data2);
  assertEqual(merged.rows.length, 2);
  assertEqual(merged.rows[0].name, 'John');
  assertEqual(merged.rows[1].name, 'Jane');
});

// Detect delimiter tests
test('detectDelimiter comma', () => {
  const csv = 'name,age,city\nJohn,30,NYC';
  assertEqual(detectDelimiter(csv), ',');
});

test('detectDelimiter semicolon', () => {
  const csv = 'name;age;city\nJohn;30;NYC';
  assertEqual(detectDelimiter(csv), ';');
});

test('detectDelimiter tab', () => {
  const csv = 'name\tage\tcity\nJohn\t30\tNYC';
  assertEqual(detectDelimiter(csv), '\t');
});

// JSON conversion tests
test('toJson', () => {
  const csv = 'name,age\nJohn,30\nJane,25';
  const data = parse(csv);
  const json = toJson(data);
  assertEqual(json.length, 2);
  assertEqual(json[0].name, 'John');
});

test('fromJson', () => {
  const json = [{ name: 'John', age: 30 }, { name: 'Jane', age: 25 }];
  const data = fromJson(json);
  assertEqual(data.headers, ['name', 'age']);
  assertEqual(data.rows.length, 2);
});

test('fromJson empty', () => {
  const json: Record<string, unknown>[] = [];
  const data = fromJson(json);
  assertEqual(data.headers, []);
  assertEqual(data.rows.length, 0);
});

// Run tests
console.log(`\n${'='.repeat(50)}`);
console.log(`Tests completed: ${testsPassed} passed, ${testsFailed} failed`);
console.log(`${'='.repeat(50)}`);

if (testsFailed > 0) {
  process.exit(1);
}

File: Perl/csv_utils/csv_utils_test.pl
#!/usr/bin/perl
# CSV Utilities Test Suite
# Comprehensive tests for AllToolkit::CSVUtils

use strict;
use warnings;
use utf8;
use FindBin;
use lib $FindBin::Bin;
require "$FindBin::Bin/mod.pl";

my $TESTS_RUN = 0;
my $TESTS_PASSED = 0;
my $TESTS_FAILED = 0;

sub test {
    my ($name, $condition) = @_;
    $TESTS_RUN++;
    if ($condition) {
        $TESTS_PASSED++;
        print "[PASS] $name\n";
    } else {
        $TESTS_FAILED++;
        print "[FAIL] $name\n";
    }
}

sub test_eq {
    my ($name, $got, $expected) = @_;
    my $pass = defined($got) && defined($expected) && $got eq $expected;
    if (!$pass) {
        $got = 'undef' unless defined $got;
        $expected = 'undef' unless defined $expected;
    }
    test("$name (got: '$got', expected: '$expected')", $pass);
}

sub test_array_eq {
    my ($name, $got, $expected) = @_;
    my $pass = 1;
    if (!defined($got) || !defined($expected)) {
        $pass = 0;
    } elsif (@$got != @$expected) {
        $pass = 0;
    } else {
        for (my $i = 0; $i < @$got; $i++) {
            if (($got->[$i] // '') ne ($expected->[$i] // '')) {
                $pass = 0;
                last;
            }
        }
    }
    test($name, $pass);
}

print "=" x 60 . "\n";
print "CSV Utilities Test Suite\n";
print "=" x 60 . "\n\n";

#==============================================================================
# Test parse_csv
#==============================================================================
print "--- Testing parse_csv ---\n";

my $result = AllToolkit::CSVUtils::parse_csv('John,30,New York');
test_array_eq("Basic parsing", $result, ['John', '30', 'New York']);

$result = AllToolkit::CSVUtils::parse_csv('"John Doe",30,"New York, NY"');
test_array_eq("Quoted fields with commas", $result, ['John Doe', '30', 'New York, NY']);

$result = AllToolkit::CSVUtils::parse_csv('"John ""Johnny"" Doe",30');
test_array_eq("Escaped quotes", $result, ['John "Johnny" Doe', '30']);

$result = AllToolkit::CSVUtils::parse_csv('');
test_array_eq("Empty string", $result, ['']);

$result = AllToolkit::CSVUtils::parse_csv(undef);
test_array_eq("Undefined input", $result, []);

$result = AllToolkit::CSVUtils::parse_csv('a,b,c', { separator => "\t" });
test_array_eq("Tab separator (no match)", $result, ['a,b,c']);

$result = AllToolkit::CSVUtils::parse_csv("a\tb\tc", { separator => "\t" });
test_array_eq("Tab separator", $result, ['a', 'b', 'c']);

#==============================================================================
# Test parse_csv_string
#==============================================================================
print "\n--- Testing parse_csv_string ---\n";

$result = AllToolkit::CSVUtils::parse_csv_string("Name,Age\nJohn,30\nJane,25");
test_eq("Multi-line parsing rows", scalar(@$result), 3);
test_array_eq("First row", $result->[0], ['Name', 'Age']);
test_array_eq("Second row", $result->[1], ['John', '30']);

$result = AllToolkit::CSVUtils::parse_csv_string("Name,Age\r\nJohn,30\r\nJane,25");
test_eq("Windows line endings rows", scalar(@$result), 3);

$result = AllToolkit::CSVUtils::parse_csv_string("");
test_array_eq("Empty string", $result, []);

$result = AllToolkit::CSVUtils::parse_csv_string(undef);
test_array_eq("Undefined input", $result, []);

#==============================================================================
# Test parse_csv_with_headers
#==============================================================================
print "\n--- Testing parse_csv_with_headers ---\n";

my $hash_result = AllToolkit::CSVUtils::parse_csv_with_headers("Name,Age,City\nJohn,30,NYC\nJane,25,LA");
test_array_eq("Headers extracted", $hash_result->{headers}, ['Name', 'Age', 'City']);
test_eq("Row count", scalar(@{$hash_result->{rows}}), 2);
test_eq("First row Name", $hash_result->{rows}[0]{Name}, 'John');
test_eq("First row Age", $hash_result->{rows}[0]{Age}, '30');
test_eq("Second row City", $hash_result->{rows}[1]{City}, 'LA');

$hash_result = AllToolkit::CSVUtils::parse_csv_with_headers("");
test_array_eq("Empty string headers", $hash_result->{headers}, []);
test_array_eq("Empty string rows", $hash_result->{rows}, []);

#==============================================================================
# Test generate_csv
#==============================================================================
print "\n--- Testing generate_csv ---\n";

my $csv_line = AllToolkit::CSVUtils::generate_csv(['John', '30', 'New York']);
test_eq("Basic generation", $csv_line, 'John,30,New York');

$csv_line = AllToolkit::CSVUtils::generate_csv(['John, Doe', '30']);
test_eq("Field with comma", $csv_line, '"John, Doe",30');

$csv_line = AllToolkit::CSVUtils::generate_csv(['John "Johnny" Doe', '30']);
test_eq("Field with quotes", $csv_line, '"John ""Johnny"" Doe",30');

$csv_line = AllToolkit::CSVUtils::generate_csv(['John', '30'], { always_quote => 1 });
test_eq("Always quote", $csv_line, '"John","30"');

$csv_line = AllToolkit::CSVUtils::generate_csv([]);
test_eq("Empty array", $csv_line, '');

#==============================================================================
# Test generate_csv_string
#==============================================================================
print "\n--- Testing generate_csv_string ---\n";

my $csv_string = AllToolkit::CSVUtils::generate_csv_string([
    ['Name', 'Age'],
    ['John', '30'],
    ['Jane', '25']
]);
test_eq("Contains newline", index($csv_string, "\n") >= 0, 1);
test_eq("Ends with newline", substr($csv_string, -1), "\n");

#==============================================================================
# Test filter_rows
#==============================================================================
print "\n--- Testing filter_rows ---\n";

my $data = [
    ['John', '30'],
    ['Jane', '25'],
    ['Bob', '35']
];

my $filtered = AllToolkit::CSVUtils::filter_rows($data, sub {
    my $row = shift;
    return ($row->[1] // 0) >= 30;
});
test_eq("Filter >= 30 count", scalar(@$filtered), 2);
test_eq("First filtered name", $filtered->[0][0], 'John');

#==============================================================================
# Test filter_rows_hash
#==============================================================================
print "\n--- Testing filter_rows_hash ---\n";

my $hash_data = {
    headers => ['Name', 'Age'],
    rows => [
        { Name => 'John', Age => '30' },
        { Name => 'Jane', Age => '25' },
        { Name => 'Bob', Age => '35' }
    ]
};

my $filtered_hash = AllToolkit::CSVUtils::filter_rows_hash($hash_data, sub {
    my $row = shift;
    return ($row->{Age} // 0) >= 30;
});
test_eq("Filter hash >= 30 count", scalar(@{$filtered_hash->{rows}}), 2);

#==============================================================================
# Test get_column
#==============================================================================
print "\n--- Testing get_column ---\n";

my $column = AllToolkit::CSVUtils::get_column($data, 0);
test_array_eq("First column", $column, ['John', 'Jane', 'Bob']);

$column = AllToolkit::CSVUtils::get_column($data, 1);
test_array_eq("Second column", $column, ['30', '25', '35']);

#==============================================================================
# Test sort_by_column
#==============================================================================
print "\n--- Testing sort_by_column ---\n";

my $sorted = AllToolkit::CSVUtils::sort_by_column($data, 1, { numeric => 1 });
test_eq("Sort by age asc first", $sorted->[0][0], 'Jane');
test_eq("Sort by age asc last", $sorted->[2][0], 'Bob');

$sorted = AllToolkit::CSVUtils::sort_by_column($data, 1, { numeric => 1, reverse => 1 });
test_eq("Sort by age desc first", $sorted->[0][0], 'Bob');

#==============================================================================
# Test distinct_values
#==============================================================================
print "\n--- Testing distinct_values ---\n";

my $distinct_data = [
    ['A', '1'],
    ['B', '2'],
    ['A', '3'],
    ['C', '1']
];

my $distinct = AllToolkit::CSVUtils::distinct_values($distinct_data, 0);
test_array_eq("Distinct first column", $distinct, ['A', 'B', 'C']);

$distinct = AllToolkit::CSVUtils::distinct_values($distinct_data, 1);
test_array_eq("Distinct second column", $distinct, ['1', '2', '3']);

#==============================================================================
# Test is_valid_csv
#==============================================================================
print "\n--- Testing is_valid_csv ---\n";

test_eq("Valid CSV", AllToolkit::CSVUtils::is_valid_csv("Name,Age\nJohn,30"), 1);
test_eq("Valid empty", AllToolkit::CSVUtils::is_valid_csv(""), 1);
test_eq("Valid undef", AllToolkit::CSVUtils::is_valid_csv(undef), 0);

#==============================================================================
# Test count_rows
#==============================================================================
print "\n--- Testing count_rows ---\n";

test_eq("Count array rows", AllToolkit::CSVUtils::count_rows($data), 3);
test_eq("Count hash rows", AllToolkit::CSVUtils::count_rows($hash_data), 3);
test_eq("Count undef", AllToolkit::CSVUtils::count_rows(undef), 0);

#==============================================================================
# Test detect_separator
#==============================================================================
print "\n--- Testing detect_separator ---\n";

test_eq("Detect comma", AllToolkit::CSVUtils::detect_separator("a,b,c"), ',');
test_eq("Detect tab", AllToolkit::CSVUtils::detect_separator("a\tb\tc"), "\t");
test_eq("Detect semicolon", AllToolkit::CSVUtils::detect_separator("a;b;c;d"), ';');

#==============================================================================
# Test slice_rows
#==============================================================================
print "\n--- Testing slice_rows ---\n";

my $sliced = AllToolkit::CSVUtils::slice_rows($data, 0, 2);
test_eq("Slice 2 rows", scalar(@$sliced), 2);
test_eq("Slice first row", $sliced->[0][0], 'John');

$sliced = AllToolkit::CSVUtils::slice_rows($data, 1, 10);
test_eq("Slice with overflow", scalar(@$sliced), 2);

#==============================================================================
# Test merge_csv
#==============================================================================
print "\n--- Testing merge_csv ---\n";

my $data1 = {
    headers => ['Name', 'Age'],
    rows => [
        { Name => 'John', Age => '30' }
    ]
};

my $data2 = {
    headers => ['Name', 'City'],
    rows => [
        { Name => 'Jane', City => 'NYC' }
    ]
};

my $merged = AllToolkit::CSVUtils::merge_csv($data1, $data2);
test_array_eq("Merged headers", $merged->{headers}, ['Name', 'Age', 'City']);
test_eq("Merged row count", scalar(@{$merged->{rows}}), 2);
test_eq("Second row has City", $merged->{rows}[1]{City}, 'NYC');

#==============================================================================
# Summary
#==============================================================================
print "\n" . "=" x 60 . "\n";
print "Test Summary\n";
print "=" x 60 . "\n";
print "Tests run:    $TESTS_RUN\n";
print "Tests passed: $TESTS_PASSED\n";
print "Tests failed: $TESTS_FAILED\n";
print "Success rate: " . sprintf("%.1f%%", ($TESTS_PASSED / $TESTS_RUN * 100)) . "\n";
print "=" x 60 . "\n";

exit($TESTS_FAILED > 0 ? 1 : 0);